providers/default/ciphers/cipher_chacha20.c

```diff
@@ -0,0 +1,292 @@
+};
+const OSSL_PARAM *chacha20_settable_ctx_params(void)
+{
+    return chacha20_known_settable_ctx_params;
+}
+
+static int chacha20_init(void *vctx, const unsigned char *key, size_t keylen,
+                         const unsigned char *iv, size_t ivlen)
+{
+    if (key != NULL) {
+        if (keylen != CHACHA20_KEYLEN) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH);
+            return 0;
+        }
+        CHACHA20_init_key(vctx, key, NULL, 1);
+    }
+
+    if (iv != NULL) {
+        if (ivlen != CHACHA20_IVLEN) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_IV_LENGTH);
+            return 0;
+        }
+        CHACHA20_init_key(vctx, NULL, iv, 0);
+    }
+    return 1;
+}
+
+static int chacha20_update(void *vctx,
+                           unsigned char *out, size_t *outl, size_t outsize,
+                           const unsigned char *in, size_t inl)
+{
+    if (outsize < inl) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
+        return 0;
+    }
+
+    if (!CHACHA20_cipher(vctx, out, in, inl)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+        return 0;
+    }
+
+    *outl = inl;
+    return 1;
+}
+
+static int chacha20_final(void *vctx, unsigned char *out, size_t *outl,
+                          size_t outsize)
+{
+    *outl = 0;
+    return 1;
+}
+
+#define IMPLEMENT_cipher_functions(nm)                                         \
+const OSSL_DISPATCH nm##_functions[] = {                                       \
+    { OSSL_FUNC_CIPHER_NEWCTX, (void (*)(void))nm##_newctx },                  \
+    { OSSL_FUNC_CIPHER_FREECTX, (void (*)(void))nm##_freectx },                \
+    { OSSL_FUNC_CIPHER_DUPCTX, (void (*)(void))nm##_dupctx },                  \
+    { OSSL_FUNC_CIPHER_ENCRYPT_INIT, (void (*)(void))nm##_einit },             \
+    { OSSL_FUNC_CIPHER_DECRYPT_INIT, (void (*)(void))nm##_dinit },             \
+    { OSSL_FUNC_CIPHER_UPDATE, (void (*)(void))nm##_update },                  \
+    { OSSL_FUNC_CIPHER_FINAL, (void (*)(void))nm##_final },                    \
+    { OSSL_FUNC_CIPHER_CIPHER, (void (*)(void))nm##_cipher},                   \
+    { OSSL_FUNC_CIPHER_GET_PARAMS, (void (*)(void))nm##_get_params },          \
+    { OSSL_FUNC_CIPHER_GETTABLE_PARAMS,(void (*)(void))nm##_gettable_params }, \
+    { OSSL_FUNC_CIPHER_GET_CTX_PARAMS, (void (*)(void))nm##_get_ctx_params },  \
+    { OSSL_FUNC_CIPHER_GETTABLE_CTX_PARAMS,                                    \
+        (void (*)(void))nm##_gettable_ctx_params },                            \
+    { OSSL_FUNC_CIPHER_SET_CTX_PARAMS, (void (*)(void))nm##_set_ctx_params },  \
+    { OSSL_FUNC_CIPHER_SETTABLE_CTX_PARAMS,                                    \
+        (void (*)(void))nm##_settable_ctx_params },                            \
+    { 0, NULL }                                                                \
+};
+
+/* chacha20_functions */
+IMPLEMENT_cipher_functions(chacha20)
```

Comment:
Not sure why we implement a macro to do this - only to call it once.

Reply:
Will we currently only implement ChaCha20, there are other variants, and may be future variants that we may want to implement.  The macro makes us prepared for such eventuality.

Reply:
I'm not keen on adding extra complexity "just in case". It would not be difficult to add a macro back in, if we subsequently needed it.

Reply:
I was thinking that I may relook at all the cipher macros after all ciphers are in and try to make the macros reusable

Reply:
For example, the "extended nonce chacha" is working its way through the CFRG right now, https://tools.ietf.org/html/draft-irtf-cfrg-xchacha-01

Reply:
I think that since the macro is in now, it's absurd to remove it just to reinsert it later. Also, with what @richsalz just said, the absurdity is getting closer to certainty.